RecoBTag/Combined/plugins/BTagProbabilityToDiscriminator.cc

```diff
@@ -0,0 +1,244 @@
+    const edm::ParameterSet &iConfig) {
+  for (auto discriminator : iConfig.getParameter<vPSet>("discriminators")) {
+    Discriminator current;
+    current.name = discriminator.getParameter<std::string>("name");
+    produces<JetTagCollection>(current.name);
+
+    for (auto intag : discriminator.getParameter<vInputTag>("numerator")) {
+      if (jet_tags_.find(intag.encode()) == jet_tags_.end()) { // new
+                                                               // probability
+        jet_tags_[intag.encode()] = consumes<JetTagCollection>(intag);
+      }
+      current.numerator.push_back(intag.encode());
+    }
+
+    for (auto intag : discriminator.getParameter<vInputTag>("denominator")) {
+      if (jet_tags_.find(intag.encode()) == jet_tags_.end()) { // new
+                                                               // probability
+        jet_tags_[intag.encode()] = consumes<JetTagCollection>(intag);
+      }
+      current.denominator.push_back(intag.encode());
+    }
+    discrims_.push_back(current);
+  }
+
+  if (jet_tags_.empty()) {
+    throw cms::Exception("RuntimeError")
+        << "The module BTagProbabilityToDiscriminator is run without any input "
+           "probability to work on!"
+        << std::endl;
+  }
+}
+
+void BTagProbabilityToDiscriminator::produce(edm::Event &iEvent,
+                                             const edm::EventSetup &iSetup) {
+  std::unordered_map<std::string, edm::Handle<JetTagCollection>>
+      tags; // caches jet tags to avoid repetitions
+  size_t size = 0;
+  bool first = true;
+  for (auto entry : jet_tags_) {
+    edm::Handle<JetTagCollection> tmp;
+    iEvent.getByToken(entry.second, tmp);
+    tags[entry.first] = tmp;
+    if (first)
+      size = tmp->size();
+    else {
+      if (tmp->size() != size) {
+        throw cms::Exception("RuntimeError")
+            << "The length of one of the input jet tag collections does not "
+               "match "
+            << "with the others, this is probably due to the probabilities "
+               "belonging to different jet collections, which is forbidden!"
+            << std::endl;
+      }
+    }
+    first = false;
+  }
+
+  // create the output collection
+  // which is a "map" RefToBase<Jet> --> float
+  vector<std::unique_ptr<JetTagCollection>> output_tags;
+  output_tags.reserve(discrims_.size());
+  for (size_t i = 0; i < discrims_.size(); ++i) {
+    output_tags.push_back(std::make_unique<JetTagCollection>(
+        *(tags.begin()->second)) // clone from the first element, will change
+                                 // the content later on
+                          );
+  }
+
+  // loop over jets
+  for (size_t idx = 0; idx < output_tags[0]->size(); idx++) {
+    auto key = output_tags[0]->key(idx); // index access should be OK as well,
+                                         // but a bit less safe, probably WAY
+                                         // faster
```

Comment:
indeed, why not count on a sane interface from JetTagCollection?

Reply:
@davidlange6 the index access assumes that all the jet tags passed follow the same sorting. Given that the probabilities should produced by the same module (now they are, cannot guarantee in future) I agree with you. I can change that if you want.